File: src/shared/utils/datatype.ts
import { DataType } from "@/shared/types/editor";

/**
 * 检查数据类型是否为基础类型（可以直接编辑值）
 */
export function isPrimitiveType(dataType: DataType): boolean {
  return [
    // 新格式（大写开头）
    "Boolean", "Int32", "Int64", "Float32", "Float64", "String",
    // 旧格式（小写，保留兼容）
    "int", "int8", "int16", "int32", "int64",
    "uint32", "uint64",
    "float", "float32", "float64",
    "bool", "string", "date", "datetime"
  ].includes(dataType as string);
}

/**
 * 检查数据类型是否为复杂类型
 */
export function isComplexType(dataType: DataType): boolean {
  return [
    "DataFrame", "Object", "Array", 
    "dataframe", "object", "array"
  ].includes(dataType as string);
}

/**
 * 获取数据类型的默认值
 */
export function getDefaultValue(dataType: DataType | string): unknown {
  const typeStr = dataType as string;
  switch (typeStr) {
    case "Boolean":
    case "bool":
      return false;
    case "Int32":
    case "Int64":
    case "int":
    case "int8":
    case "int16":
    case "int32":
    case "int64":
    case "uint32":
    case "uint64":
      return 0;
    case "Float32":
    case "Float64":
    case "float":
    case "float32":
    case "float64":
      return 0.0;
    case "String":
    case "string":
      return "";
    case "Array":
    case "array":
      return [];
    case "Object":
    case "object":
      return {};
    default:
      return undefined;
  }
}

/**
 * 将旧格式的类型名转换为新格式
 */
export function normalizeDataType(dataType: string): DataType {
  const typeMap: Record<string, DataType> = {
    "bool": "Boolean",
    "int": "Int32",
    "int8": "Int32",
    "int16": "Int32",
    "int32": "Int32",
    "int64": "Int64",
    "uint32": "Int32",
    "uint64": "Int64",
    "float": "Float32",
    "float32": "Float32",
    "float64": "Float64",
    "string": "String",
    "object": "Object",
    "array": "Array",
    "dataframe": "DataFrame",
  };
  
  return (typeMap[dataType.toLowerCase()] || dataType) as DataType;
}

File: src-tauri/crates/yss-tabular-polars/src/data_type.rs
//! Canonical database dtype names at the Polars boundary.

use polars::prelude::{DataType, TimeUnit};

pub fn dtype_from_string(s: &str) -> Result<DataType, String> {
    let dtype = match s.to_lowercase().as_str() {
        "int8" | "i8" => DataType::Int8,
        "int16" | "i16" => DataType::Int16,
        "int32" | "i32" => DataType::Int32,
        "int64" | "i64" => DataType::Int64,
        "uint8" | "u8" => DataType::UInt8,
        "uint16" | "u16" => DataType::UInt16,
        "uint32" | "u32" => DataType::UInt32,
        "uint64" | "u64" => DataType::UInt64,
        "float32" | "f32" => DataType::Float32,
        "float64" | "f64" => DataType::Float64,
        "bool" | "boolean" => DataType::Boolean,
        "date" => DataType::Date,
        "datetime" | "dt" => DataType::Datetime(TimeUnit::Microseconds, None),
        "string" | "str" | "utf8" => DataType::String,
        "categorical" | "category" | "cat" => {
            use polars_dtype::categorical::Categories;
            DataType::from_categories(Categories::global())
        }
        _ => return Err(format!("Unknown database dtype '{s}'")),
    };
    Ok(dtype)
}

pub fn dtype_to_string(dt: &DataType) -> Result<String, String> {
    let dtype = match dt {
        DataType::Int8 => "Int8".into(),
        DataType::Int16 => "Int16".into(),
        DataType::Int32 => "Int32".into(),
        DataType::Int64 => "Int64".into(),
        DataType::UInt8 => "UInt8".into(),
        DataType::UInt16 => "UInt16".into(),
        DataType::UInt32 => "UInt32".into(),
        DataType::UInt64 => "UInt64".into(),
        DataType::Float32 => "Float32".into(),
        DataType::Float64 => "Float64".into(),
        DataType::Boolean => "Boolean".into(),
        DataType::String => "String".into(),
        DataType::Categorical(_, _) => "Categorical".into(),
        DataType::Date => "Date".into(),
        DataType::Datetime(_, _) => "DateTime".into(),
        _ => return Err(format!("Unsupported database dtype {dt:?}")),
    };
    Ok(dtype)
}
